File: 6.z-字形变换.js
/*
 * @lc app=leetcode.cn id=6 lang=javascript
 *
 * [6] Z 字形变换
 */

// @lc code=start
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function (s, numRows) {
  if (numRows === 1) {
    return s;
  }
  let len = Math.min(s.length, numRows);
  let rows = new Array(len).fill("");
  let loc = 0;
  let step = -1;
  for (let c of s) {
    rows[loc] += c;
    if (loc === 0 || loc === numRows - 1) {
      step = -step;
    }
    loc += step;
  }
  return rows.join("");
};
// @lc code=end

File: 6.z-字形变换.rs
/*
 * @lc app=leetcode.cn id=6 lang=rust
 *
 * [6] Z 字形变换
 */

// @lc code=start
impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        // 为什么输入类型总是i32
        let num_rows = num_rows as usize;
        // 每行一个String
        let mut rows = vec![String::new(); num_rows];
        // z字形往复的迭代器，01232101232......
        let iter = (0..num_rows).chain((1..num_rows - 1).rev()).cycle();
        // 按迭代器给出的下标访问对应行，推入字符
        iter.zip(s.chars()).for_each(|(i, c)| rows[i].push(c));
        // collect连接每行
        rows.into_iter().collect()
    }
}
// @lc code=end
